add THREADING_LAYER_PRIORITY & NUMBA_THREADING_LAYER_PRIORITY (#7234)
numba/numba: docs/source/user/threading-layer.rst

safe under various forms of parallel execution, the second is through explicit
selection via the threading layer name (e.g. ``tbb``).

> Setting the threading layer priority
> ------------------------------------

Reviewer: ```suggestion
Setting the threading layer selection priority
----------------------------------------------
```

> By default, The threading layers are searched in the order of

Reviewer: ```suggestion
By default, the threading layers are searched in the order of
```

Author: GitHub marked this as outdated. You already made a suggestion in b11cf88d0e5581db9436bcda8ae875d29cedf0c3.

'tbb', 'omp', and 'workqueue'.

> If you want to let Numba choice the threading layer

Reviewer: ```suggestion
If you want to let Numba choose the threading layer
```

Author: GitHub marked this as outdated. You already made a suggestion in b11cf88.

> By default, The threading layers are searched in the order of
> 'tbb', 'omp', and 'workqueue'.
> If you want to let Numba choice the threading layer
> based on availability automatically but with a different priority,
> it can be changed via the environment variable
> ``NUMBA_THREADING_LAYER_PRIORITY`` or through assignment to
> ``numba.config.THREADING_LAYER_PRIORITY``.
> 
> For example, if you want Numba to choose ``omp`` first if available,
> then ``tbb`` and so on, you can set either
> ``NUMBA_THREADING_LAYER_PRIORITY="omp tbb workqueue"`` or
> ``numba.config.THREADING_LAYER_PRIORITY = ["omp", "tbb", "workqueue"]``.

Reviewer: ```suggestion
By default the threading layers are searched in the order of ``'tbb'``,
``'omp'``, then ``'workqueue'``. To change this search order whilst
maintaining the selection of a threading layer based on availability, the
environment variable :envvar:`NUMBA_THREADING_LAYER_PRIORITY` can be used.

For example, to instruct Numba to choose ``omp`` first if available,
then ``tbb`` and so on, set the environment variable as
``NUMBA_THREADING_LAYER_PRIORITY="omp tbb workqueue"``.
```

I suggest not mentioning assignment to the config variable as it leads to confusion in practice as it's mutating a global that could well have already been read (e.g. if a user specifies a signature to a module level `@njit(parallel=True)` function.

Reviewer: I think with the noted addition of similar caveats about setting it ahead of compilation this is ok to add.

> ``numba.config.THREADING_LAYER_PRIORITY = ["omp", "tbb", "workqueue"]``.

Reviewer: ```suggestion
For example, if you want Numba to choose ``omp`` first if available,
then ``tbb`` and so on, you can set either
``NUMBA_THREADING_LAYER_PRIORITY="omp tbb workqueue"`` or
``numba.config.THREADING_LAYER_PRIORITY = ["omp", "tbb", "workqueue"]``.

If you want to let Numba choice the threading layer
based on availability automatically but with a different priority,
it can be changed via the environment variable
``NUMBA_THREADING_LAYER_PRIORITY`` or through assignment to
``numba.config.THREADING_LAYER_PRIORITY``.
```
I suggest reordering these such that the common use example (env var definition) is first.

Reviewer: Actually, it seems ok the original way around on a re-read.
